### PR TITLE
bitstream: fixes

### PR DESCRIPTION
**Pull request by mildsunrise**
Fixes several issues in bitstream.h (BitReader and BitWriter):

 - in C++, [it is undefined behavior to do `dword >> 32` or `dword << 32`](https://stackoverflow.com/a/18918340/710951). And we are doing that in several points of the code. This is one of the UBs that hardly leads to crashes, but it did lead to an actual bug, which is observed in #253 (see description). In `Get()`, when we combine cache with ret:

   https://github.com/medooze/media-server/blob/6b369a170791bf25f80b056e6f83855e5a5e6291/include/bitstream.h#L91

   This assumes that the lower bits of `cache` are zero. Which is guaranteed since `SkipCached()` does a right shift:

   https://github.com/medooze/media-server/blob/6b369a170791bf25f80b056e6f83855e5a5e6291/include/bitstream.h#L268

   ...except when `n` is 32, in which case this invokes UB. In most implementations / architectures this will result in leaving `cache` unchanged rather than zeroed.

 - `GetVariableLengthUnsigned` makes no sense. It is an AV1 primitive (essentially identical to H.264 `ue(v)`) and it should be implemented as: 
Screenshot from 2024-02-22 21-29-42

 - `Get()` doesn't set error condition if there are not enough bits to read (or rather, it only does that in specific cases).

### CODE REVIEW

**Comment by mildsunrise** [comment]
More issues that I have:

- `GolombExpDecoder::Decode()` invokes UB with certain user input.
- `BitWritter::Put()` does not signal error if there isn't enough space in buffer (at least, not immediately).
- It is possible to underflow `cached`, for example by reading more bits than present. this can then cause `Left()` to underflow as well.

File: include/bitstream.h [hunk 1]
```diff
@@ -91,6 +91,9 @@ class BitReader
 			ret = cache >> (32-n);
 			//Cache next
 			Cache();
+			//Check available
+			if (cached<a)
```
Comment (murillo128) [inline include/bitstream.h:95]: shouldn't this check be done inside `GetChached()` ?

File: include/bitstream.h [hunk 2]
```diff
@@ -265,7 +270,10 @@ class BitReader
 		//Check length
 		if (!n) return;
 		//Move
-		cache = cache << n;
+		if (n < 32)
+			cache = cache << n;
+		else
+			cache = 0;
 		//Update cached bytes
 		cached -= n;
```
Comment (murillo128) [inline include/bitstream.h:278]: shouldn't we check that `n <= cached`?

File: include/bitstream.h [hunk 3]
```diff
@@ -414,6 +422,8 @@ class BitWritter{
 			//Increase cached
 			cached = b;
 		} else {
+			//Don't invoke UB if empty
+			if (!n) return v;
```
Comment (murillo128) [inline include/bitstream.h:426]: should we put this check on the else?

File: include/bitstream.h [hunk 4]
```diff
@@ -414,6 +422,8 @@ class BitWritter{
 			//Increase cached
 			cached = b;
 		} else {
```
Comment (murillo128) [inline include/bitstream.h:424]:
```suggestion
		} else if (n) {
```

**Review by murillo128 — CHANGES_REQUESTED** [review]

File: include/bitstream.h (hunk 1 above)
Comment (mildsunrise) [reply include/bitstream.h:95]: `GetCached()` is called from two places, and only this call site needs the check. but if you prefer to move it there for clarity, we can do that too

File: include/bitstream.h (hunk 2 above)
Comment (mildsunrise) [reply include/bitstream.h:278]: this is the issue I pointed above (It is possible to underflow cached, for example by reading more bits than present), so yeah let's fix that too

File: include/bitstream.h (hunk 3 above)
Comment (mildsunrise) [reply include/bitstream.h:426]: the idea was to skip the FlushCache() call afterwards, but yes, we can also do that if you prefer

File: include/bitstream.h (hunk 2 above)
Comment (mildsunrise) [reply include/bitstream.h]: applied

File: include/bitstream.h (hunk 1 above)
Comment (murillo128) [reply include/bitstream.h:95]: better do it in the inner most layer always

File: include/bitstream.h (hunk 3 above)
Comment (murillo128) [reply include/bitstream.h:428]: we can put it at the begginging of the function, or as a first clause of the if-else-if

**Review by murillo128 — APPROVED** [review]
lgtm with just to minor comments, feel free to merge it when you are confortable with it

**Comment by bcostdolby** [comment]
Overall this class makes some basic assumptions on its usage that are not enforced which seems to allow these kinds of issues (especially if needing to decode some kind of variable amount of bits).

It seems there is an "error" state on the class which seems to be its way of notifying issues to the caller. Should we also set that in all these cases (in addition to preventing UB)?

I would think we need to enforce the value passed to this is always <=32 in general for all the public functions.

For example even the Get() does a:
ret = cache >> (32 - n)

if n is given as larger than 32 this is also UB.


I would suggest a common check maybe for all public functions (seems they are all public) for places like:
* Get(n) : Where n > 32
* Check(n) : Where n > 32 (I guess covered by the get, but still)
and others...:
* Skip(n)
* Peek(n)
* SkipCached() 
* GetCached()

seems we are good at checking n == 0 but not n > 32.

**Comment by bcostdolby** [comment]
Also GetNonSymmetric() could probably use some comments describing what it is trying to do. The name doesnt really tell us much.

**Review by bcostdolby — CHANGES_REQUESTED** [review]
I added comments to overall ticket, this class needs more changes I think. Actually reading through the comments, it seems Alba identified some of these but maybe they were not pushed?

**Comment by mildsunrise** [comment]
many of the tasks described in the comments of this PR, as well as outstanding comments in diffs, have been addressed now.

I'll make a list of the ones that haven't been addressed yet (please everyone verify I'm not missing anything):

- GolombExpDecoder::Decode() invokes UB with certain user input
- BitWritter::Put() does not signal error if there isn't enough space in buffer (at least, not immediately)
- check for n>32 in Check() too, requested by @bcostdolby (actually I'd change all these checks into asserts :smiling_imp: but I understand if others don't wanna)
- GetNonSymmetric() could probably use some comments describing what it is trying to do

some questions:

> It seems there is an "error" state on the class which seems to be its way of notifying issues to the caller. Should we also set that in all these cases (in addition to preventing UB)?

regarding this, I think we should do the opposite, and convert this file to use exceptions. not sure how others see it

in any case, I'd merge this now and we can open a new PR for further fixes.
also, @bcostdolby you mentioned you added things to "overall ticket", which ticket is that?

**Review by murillo128 — APPROVED** [review]
lgtm, let's do the rest of the fixes in a different PR. Probably also introduce exceptions.

**Comment by murillo128** [comment]
+1 to replace error flags by proper exception handling.

Not sure about asserts, I would be worried about crashing on production
because a software bug.

**Pull request merged**